### ISSUE TITLE
Error in loss scale with Onebitadam

### ISSUE THREAD

**Issue by ShivanshuPurohit** [microsoft/DeepSpeed#753]
I was trying to train a gpt-2 like model, and with onebitadam, I got stuck in an infinite loop like so
```
10.140.11.15: Grad overflow on iteration 139
10.140.11.15: [2021-02-12 11:05:20,845] [INFO] [fused_optimizer.py:315:_update_scale] Reducing dynamic loss scale from 3.2311742677852644e-27 to 1.6155871338926322e-27
10.140.11.15: [2021-02-12 11:05:20,845] [INFO] [logging.py:60:log_dist] [Rank 0] Overflow detected. Skipping step. Attempted loss scale: 3.2311742677852644e-27, reducing to 1.6155871338926322e-27
```
It should've scaled the loss to 32784, like adam does, and then the training woul've resumed. But it just keeps halving the loss scale ad infinitum.

**Comment by awan-10** [comment]
@ShivanshuPurohit -- Can you please let us know what is the freeze_step argument? Onebit Adam needs a warmup phase specified by freeze_step. Increase the freeze_step to a larger number and try again. It needs to be determined experimentally. We only had it for BERT pretraining in the example code. For other model sizes and structures, it may need to be increased.

cc @conglongli -- please add if I missed something here.

**Comment by conglongli** [comment]
@ShivanshuPurohit The two parameters you need to be careful is freeze_step and cuda_aware. You should only set cuda_aware to True if you use InfiniBand interconnect and a CUDA-Aware MPI library. For freeze_step, our experience on BERT pre-train and Squad fine-tune is that you have to set it at around 15-20% of the total training steps. Please read https://www.deepspeed.ai/tutorials/onebit-adam/ and https://arxiv.org/abs/2102.02888 for more details.

**Comment by ShivanshuPurohit** [comment]
`"freeze_step": 400` and `"cuda_aware": true` in my config. I didn't know freeze_step needed to be this high.

**Comment by ShivanshuPurohit** [comment]
So I changed the freeze_step to 40000 (total iters is 320,000), but even then, this is what I'm getting
```
 iteration      688/  320000 | elapsed time per iteration (ms): 2597.9 | learning rate: 1.000E-03 | lm loss: 0.000000E+00 | loss scale: 1.0 | number of skipped iterations:   1 | number of nan iterations:   1 |
time (ms) | forward: 0.00 | backward: 0.00 | optimizer: 0.00 | batch generator: 0.00
[2021-02-13 08:49:26,860] [INFO] [fused_optimizer.py:314:_update_scale]
Grad overflow on iteration 688
[2021-02-13 08:49:26,860] [INFO] [fused_optimizer.py:314:_update_scale]
Grad overflow on iteration 688
[2021-02-13 08:49:26,860] [INFO] [fused_optimizer.py:314:_update_scale]
Grad overflow on iteration 688
[2021-02-13 08:49:26,860] [INFO] [fused_optimizer.py:314:_update_scale]
Grad overflow on iteration 688
[2021-02-13 08:49:26,860] [INFO] [fused_optimizer.py:314:_update_scale]
Grad overflow on iteration 688
[2021-02-13 08:49:26,860] [INFO] [fused_optimizer.py:314:_update_scale]
Grad overflow on iteration 688
[2021-02-13 08:49:26,860] [INFO] [fused_optimizer.py:315:_update_scale] Reducing dynamic loss scale from 1 to 1
[2021-02-13 08:49:26,860] [INFO] [fused_optimizer.py:314:_update_scale]
Grad overflow on iteration 688
[2021-02-13 08:49:26,860] [INFO] [fused_optimizer.py:315:_update_scale] Reducing dynamic loss scale from 1 to 1
[2021-02-13 08:49:26,860] [INFO] [fused_optimizer.py:315:_update_scale] Reducing dynamic loss scale from 1 to 1
[2021-02-13 08:49:26,860] [INFO] [fused_optimizer.py:315:_update_scale] Reducing dynamic loss scale from 1 to 1
[2021-02-13 08:49:26,860] [INFO] [fused_optimizer.py:315:_update_scale] Reducing dynamic loss scale from 1 to 1
[2021-02-13 08:49:26,860] [INFO] [fused_optimizer.py:315:_update_scale] Reducing dynamic loss scale from 1 to 1
[2021-02-13 08:49:26,860] [INFO] [fused_optimizer.py:314:_update_scale]
Grad overflow on iteration 688
[2021-02-13 08:49:26,860] [INFO] [fused_optimizer.py:315:_update_scale] Reducing dynamic loss scale from 1 to 1
[2021-02-13 08:49:26,860] [INFO] [fused_optimizer.py:315:_update_scale] Reducing dynamic loss scale from 1 to 1
[2021-02-13 08:49:26,860] [INFO] [logging.py:60:log_dist] [Rank 0] Overflow detected. Skipping step. Attempted loss scale: 1, reducing to 1
rank=0 time (ms) | compute_norm: 0.00 | overflow_check: 1.79
```
Rinse and repeat. The loss scale isn't going below 1 this time, but nevertheless, no learning is taking place.
```
 iteration      700/  320000 | elapsed time per iteration (ms): 2605.1 | learning rate: 1.000E-03 | lm loss: 0.000000E+00 | loss scale: 1.0 | number of skipped iterations:   1 | number of nan iterations:   1 |
time (ms) | forward: 0.00 | backward: 0.00 | optimizer: 0.00 | batch generator: 0.00
--------------------------------------------------------------------------------------
 validation loss at iteration 700 | lm loss value: NAN | lm loss PPL: 4.851652E+08 |
--------------------------------------------------------------------------------------
```
Like so

**Comment by conglongli** [comment]
@ShivanshuPurohit So you mean that the unexpected overflow happened before reaching the freeze_step number of steps? I notice that the learning rate is 1.000E-03. Is this the same lr you used for baseline Adam? If not, this overflow might be caused by too large learning rate. Please make sure to use the same training parameters for adam and 1-bit adam (such as lr, lr_schedule, lr_offset, and everything related to learning rate schedule), because 1-bit adam is designed to achieve the same convergence when using the same training parameters as adam.

**Comment by ShivanshuPurohit** [comment]
I am using
```
"scheduler": {
 "type": "WarmupDecayLR",
  "params": {
    "warmup_min_lr": 0,
    "warmup_max_lr": 0.0002,
    "warmup_num_steps": 1000,
    "total_num_steps": 300000
    }
  }
 }
```
In the onebitadam config,  and 
```
# GPT options:
NLAYERS=12
NHIDDEN=1024
NHEADS=16
SEQLEN=1024
LR="2.0e-4"
MINLR="2.0e-5"
WEIGHTDECAY=0
DROPOUT=0
SPARSITY='none'
TRAIN_ITERS=320000
```
In the training script. I don't understand how its reaching 1e-3 lr when the scheduler maxes out on 2e-4

**Comment by conglongli** [comment]
This looks like an incorrect setting of learning rate offset (an absolute value added to all learning rates). Check if you have something like "--lr_offset" flag in your ds_train script. For some of our examples for Lamb optizimer, we used a lr_offset = 10e-4 like here https://github.com/microsoft/DeepSpeedExamples/blob/master/bing_bert/ds_train_bert_bsz64k_seq128.sh. But for adam and 1-bit adam you should have a zero lr_offset unless you set it specifically.

**Comment by ShivanshuPurohit** [comment]
You were right. The lr was 0.001 in the config. But even now, when the training starts, the lr is 0.0001 from the get go, instead of warming up as the sceduler defines.

**Comment by ShivanshuPurohit** [comment]
The issue is resolved. Turns out the lr mentioned in the 1-bit adam config was being used for training rather than the one defined in the script. Thanks.
